File: tosctl/src/validator-auth-native/src/certificate_proof.rs
use crate::{
    committee::ChainContext, committee_proof::verify_committee_proof, proof::verify_native_response,
};
use tos_validator_auth::{
    api_semantics::{api_request_id, validate_api_request},
    client::ResponseVerifier,
    codec::{decode, encode, Error, Hash},
    crypto::object_id,
    transfer::ObjectReader,
    types::*,
    verify::VerifiedCertificate,
};

pub struct VerifiedNativeCertificate {
    anchor: Anchor,
    certificate: VerifiedCertificate,
}

/// One call owns its admission result. There is no cross-request verdict cache.
pub struct PreparedNativeCertificate {
    request_id: Hash,
    anchor: Anchor,
    chain: ChainContext,
    expected: Duty,
    verified: Option<VerifiedNativeCertificate>,
}
/// The application supplies independently trusted finality and native duty data.
pub struct NativeCertificateVerifier {
    pub anchor: Anchor,
    pub chain: ChainContext,
    pub expected: Duty,
}
impl ResponseVerifier for NativeCertificateVerifier {
    type Prepared = PreparedNativeCertificate;
    type Output = VerifiedNativeCertificate;
    fn prepare<F: FnMut(&ObjectRef, u8) -> Result<Vec<u8>, Error>>(
        &self,
        method: u8,
        request: &[u8],
        reader: &mut ObjectReader<F>,
    ) -> Result<Self::Prepared, Error> {
        admit_bytes(request, &[])?;
        let verified = match method {
            12 => {
                validate_api_request(method, request, reader)?;
                if decode::<GetCertificateRequest>(request)?.anchor != self.anchor {
                    return Err(Error("certificate-anchor"));
                }
                None
            }
            13 => Some(verify_native_certificate(
                &decode(request)?,
                &self.anchor,
                &self.chain,
                &self.expected,
                reader,
            )?),
            _ => return Err(Error("unsupported-native-method")),
        };
        Ok(PreparedNativeCertificate {
            request_id: api_request_id(method, request)?,
            anchor: self.anchor.clone(),
            chain: self.chain.clone(),
            expected: self.expected.clone(),
            verified,
        })
    }
    fn verify<F: FnMut(&ObjectRef, u8) -> Result<Vec<u8>, Error>>(
        &self,
        prepared: Self::Prepared,
        method: u8,
        id: Hash,
        request: &[u8],
        response: &[u8],
        reader: &mut ObjectReader<F>,
    ) -> Result<Self::Output, Error> {
        admit_bytes(request, response)?;
        if prepared.request_id != id
            || api_request_id(method, request)? != id
            || prepared.anchor != self.anchor
            || prepared.chain != self.chain
            || prepared.expected != self.expected
        {
            return Err(Error("certificate-prepared-context"));
        }
        match (method, prepared.verified) {
            (13, Some(verified)) => {
                if decode::<VerifyResult>(response)? != verified.result()? {
                    return Err(Error("verified-result"));
                }
                Ok(verified)
            }
            (12, None) => verify_native_certificate_response(
                method,
                request,
                response,
                &self.anchor,
                &self.chain,
                &self.expected,
                reader,
            ),
            _ => Err(Error("certificate-prepared-context")),
        }
    }
}
impl VerifiedNativeCertificate {
    pub fn anchor(&self) -> &Anchor {
        &self.anchor
    }
    pub fn certificate(&self) -> &VerifiedCertificate {
        &self.certificate
    }
    pub fn result(&self) -> Result<VerifyResult, Error> {
        let duty = self.certificate.duty();
        Ok(VerifyResult {
            anchor: self.anchor.clone(),
            certificate_id: *self.certificate.certificate_id(),
            policy: duty.policy,
            committee: duty.committee,
            duty: object_id("duty", duty)?,
            signers: self.certificate.signers().to_vec(),
            weight: self.certificate.weight(),
        })
    }
}
/// Finality and the exact native duty are independently pinned by the caller.
/// Peer certificates cannot choose session birth, expected duty or the anchor.
pub fn verify_native_certificate<F: FnMut(&ObjectRef, u8) -> Result<Vec<u8>, Error>>(
    request: &VerifyCertificateRequest,
    anchor: &Anchor,
    chain: &ChainContext,
    expected: &Duty,
    reader: &mut ObjectReader<F>,
) -> Result<VerifiedNativeCertificate, Error> {
    if request.anchor != *anchor {
        return Err(Error("certificate-anchor"));
    }
    if expected.network != chain.network
        || expected.genesis_root != chain.genesis_root
        || expected.genesis_file != chain.genesis_file
        || expected.anchor_mc != anchor.seqno
    {
        return Err(Error("certificate-chain-context"));
    }
    let certificate = decode::<Certificate>(&reader.resolve(&request.certificate, 4)?)?;
    if certificate.duty != *expected {
        return Err(Error("expected-context"));
    }
    let committee = verify_committee_proof(
        &request.committee,
        anchor,
        chain,
        expected.workchain,
        expected.shard,
        expected.catchain,
        reader,
    )?;
    let snapshot = committee.snapshot();
    verify_native_response(
        9,
        &encode(&GetPolicyRequest { anchor: anchor.clone(), policy_id: expected.policy })?,
        &encode(&PolicyResult {
            anchor: anchor.clone(),
            policy: snapshot.policy().clone(),
            proof: request.policy.clone(),
        })?,
        anchor,
        chain.network,
        reader,
    )?;
    let certificate = snapshot.verify_certificate(&certificate, expected)?;
    Ok(VerifiedNativeCertificate { anchor: anchor.clone(), certificate })
}

pub fn verify_native_certificate_response<F: FnMut(&ObjectRef, u8) -> Result<Vec<u8>, Error>>(
    method: u8,
    request: &[u8],
    response: &[u8],
    anchor: &Anchor,
    chain: &ChainContext,
    expected: &Duty,
    reader: &mut ObjectReader<F>,
) -> Result<VerifiedNativeCertificate, Error> {
    admit_bytes(request, response)?;
    match method {
        12 => {
            let q = decode::<GetCertificateRequest>(request)?;
            let r = decode::<CertificateResult>(response)?;
            if q.anchor != *anchor || r.anchor != *anchor {
                return Err(Error("certificate-anchor"));
            }
            if r.era != 1 || r.interface_digest != INTERFACE_FINGERPRINT {
                return Err(Error("certificate-era"));
            }
            let verified = verify_native_certificate(
                &VerifyCertificateRequest {
                    anchor: anchor.clone(),
                    certificate: r.certificate,
                    committee: r.committee,
                    policy: r.policy,
                },
                anchor,
                chain,
                expected,
                reader,
            )?;
            if *verified.certificate().certificate_id() != q.certificate_id {
                return Err(Error("certificate-id"));
            }
            Ok(verified)
        }
        13 => {
            let q = decode::<VerifyCertificateRequest>(request)?;
            let r = decode::<VerifyResult>(response)?;
            let verified = verify_native_certificate(&q, anchor, chain, expected, reader)?;
            if r != verified.result()? {
                return Err(Error("verified-result"));
            }
            Ok(verified)
        }
        _ => Err(Error("unsupported-native-method")),
    }
}

fn admit_bytes(request: &[u8], response: &[u8]) -> Result<(), Error> {
    if request.len() > 2_000_000 || response.len() > 2_000_000 {
        return Err(Error("api-binary-bound"));
    }
    Ok(())
}

File: validator/auth/certificate-proof.cpp
#include "certificate-proof.h"
namespace tos::auth {
Result<VerifyResult> VerifiedNativeCertificate::result() const {
  const auto& duty = certificate_.duty();
  auto id = object_id("duty", duty);
  if (!id.ok())
    return id.error();
  return VerifyResult{anchor_,    certificate_.certificate_id(), duty.policy_,         duty.committee_,
                      id.value(), certificate_.signers(),        certificate_.weight()};
}
Result<VerifiedNativeCertificate> verify_native_certificate(const VerifyCertificateRequest& request,
                                                            const Anchor& anchor, const ChainContext& chain,
                                                            const Duty& expected, ObjectReader& reader) {
  return verify_native_certificate(request, anchor, chain,
                                   NativeDutyResolver([&](const Duty&) -> Result<Duty> { return expected; }), reader);
}
Result<VerifiedNativeCertificate> verify_native_certificate(const VerifyCertificateRequest& request,
                                                            const Anchor& anchor, const ChainContext& chain,
                                                            const NativeDutyResolver& resolve, ObjectReader& reader) {
  if (request.anchor_ != anchor)
    return Error{"certificate-anchor"};
  auto raw = reader.resolve(request.certificate_, 4);
  if (!raw.ok())
    return raw.error();
  auto cert = decode<Certificate>(raw.value());
  if (!cert.ok())
    return cert.error();
  if (!resolve)
    return Error{"native-duty-source"};
  auto resolved = resolve(cert.value().duty_);
  if (!resolved.ok())
    return resolved.error();
  const auto& expected = resolved.value();
  if (expected.network_ != chain.network || expected.genesis_root_ != chain.genesis_root ||
      expected.genesis_file_ != chain.genesis_file || expected.anchor_mc_ != anchor.seqno_)
    return Error{"certificate-chain-context"};
  if (cert.value().duty_ != expected)
    return Error{"expected-context"};
  auto committee = verify_committee_proof(request.committee_, anchor, chain, {expected.workchain_, expected.shard_},
                                          expected.catchain_, reader);
  if (!committee.ok())
    return committee.error();
  const auto& snapshot = committee.value().snapshot();
  auto query = encode(GetPolicyRequest{anchor, expected.policy_});
  if (!query.ok())
    return query.error();
  auto response = encode(PolicyResult{anchor, snapshot.policy(), request.policy_});
  if (!response.ok())
    return response.error();
  auto policy = verify_native_response(9, query.value(), response.value(), anchor, chain.network, reader);
  if (!policy.ok())
    return policy.error();
  auto verified = snapshot.verify(cert.value(), expected);
  if (!verified.ok())
    return verified.error();
  return VerifiedNativeCertificate(anchor, std::move(verified.value()));
}
Result<VerifiedNativeCertificate> verify_native_certificate_response(std::uint8_t method,
                                                                     std::span<const std::uint8_t> request,
                                                                     std::span<const std::uint8_t> response,
                                                                     const Anchor& anchor, const ChainContext& chain,
                                                                     const Duty& expected, ObjectReader& reader) {
  if (request.size() > 2000000 || response.size() > 2000000)
    return Error{"api-binary-bound"};
  if (method == 12) {
    auto q = decode<GetCertificateRequest>(request);
    if (!q.ok())
      return q.error();
    auto r = decode<CertificateResult>(response);
    if (!r.ok())
      return r.error();
    if (q.value().anchor_ != anchor || r.value().anchor_ != anchor)
      return Error{"certificate-anchor"};
    if (r.value().era_ != 1 || r.value().interface_digest_ != interface_fingerprint)
      return Error{"certificate-era"};
    auto verified = verify_native_certificate({anchor, r.value().certificate_, r.value().committee_, r.value().policy_},
                                              anchor, chain, expected, reader);
    if (!verified.ok())
      return verified.error();
    if (verified.value().certificate().certificate_id() != q.value().certificate_id_)
      return Error{"certificate-id"};
    return verified;
  }
  if (method == 13) {
    auto q = decode<VerifyCertificateRequest>(request);
    if (!q.ok())
      return q.error();
    auto r = decode<VerifyResult>(response);
    if (!r.ok())
      return r.error();
    auto verified = verify_native_certificate(q.value(), anchor, chain, expected, reader);
    if (!verified.ok())
      return verified.error();
    auto actual = verified.value().result();
    if (!actual.ok())
      return actual.error();
    if (r.value() != actual.value())
      return Error{"verified-result"};
    return verified;
  }
  return Error{"unsupported-native-method"};
}
}  // namespace tos::auth
